src/cauchy.jl

```diff
@@ -1,54 +1,72 @@
 #### Cauchy matrix
 
 export Cauchy
+
 """
-[`Cauchy` matrix](http://en.wikipedia.org/wiki/Cauchy_matrix)
+    [`Cauchy` matrix](http://en.wikipedia.org/wiki/Cauchy_matrix)
+
+* `Cauchy(x,y)[i,j] = 1 / (x[i] + y[j])`
+* `Cauchy(x) = Cauchy(x,x)`
+* `Cauchy(k::Int) = Cauchy(1:k)`
+
+Both `x` and `y` can be any iterable (typically vectors),
+but all elements of `x` must have the same type; likewise for `y`.
+
 ```julia
-Cauchy(x,y)[i,j]=1/(x[i]+y[j])
-Cauchy(x)=Cauchy(x,x)
-cauchy(k::Number)=Cauchy(collect(1:k))
-
-julia> Cauchy([1,2,3],[3,4,5])
-3x3 Cauchy{Int64}:
- 0.25      0.2       0.166667
- 0.2       0.166667  0.142857
- 0.166667  0.142857  0.125
-
-julia> Cauchy([1,2,3])
-3x3 Cauchy{Int64}:
- 0.5       0.333333  0.25
- 0.333333  0.25      0.2
- 0.25      0.2       0.166667
-
-julia> Cauchy(pi)
-3x3 Cauchy{Float64}:
- 0.5       0.333333  0.25
- 0.333333  0.25      0.2
- 0.25      0.2       0.166667
+julia> Cauchy([2.0 1], (0, 1, 2))
+2×3 Cauchy{Float64, Matrix{Float64}, Tuple{Int64, Int64, Int64}}:
+ 0.5  0.333333  0.25
+ 1.0  0.5       0.333333
+
+julia> Cauchy(1:3)
+3×3 Cauchy{Rational{Int64}, UnitRange{Int64}, UnitRange{Int64}}:
+ 1//2  1//3  1//4
+ 1//3  1//4  1//5
+ 1//4  1//5  1//6
+
+julia> Cauchy(3)
+3×3 Cauchy{Rational{Int64}, UnitRange{Int64}, UnitRange{Int64}}:
+ 1//2  1//3  1//4
+ 1//3  1//4  1//5
+ 1//4  1//5  1//6
 ```
 """
-struct Cauchy{T} <: AbstractMatrix{T}
-    x::Vector{T} #
-    y::Vector{T} #
-end # immutable
-
-function Cauchy(k::Number)
-         Cauchy(collect(1:k),collect(1:k))
-end
+struct Cauchy{T,X,Y} <: AbstractMatrix{T}
+    x::X
+    y::Y
 
-function Cauchy(x::Vector)
-         Cauchy(x,x)
+    # The element type T corresponds to the reciprocal
+    # of the sum of pairs of elements of x and y
+    function Cauchy(x::X, y::Y) where {X, Y}
+        Tx = eltype(first(x))
+        Ty = eltype(first(y))
+        all(==(Tx), eltype.(x)) || throw(ArgumentError("inconsistent x element types"))
+        all(==(Ty), eltype.(y)) || throw(ArgumentError("inconsistent y element types"))
```

Comment:
Are `x` and `y` more abstract than an `AbstractArray{T}`? Could we enforce consistency in another way?

Reply:
AFAIK, in every other special matrix in this package, the input types (when relevant) were originally just `Vector` and we have generalized many of them to `AbstractVector` so that users can use ranges like `1:3`.
For reasons unknown to me, this particular special matrix allowed more general iterable inputs, like a tuple of the form `(1.0, 1//2, 3f0)`.  The test case even had a row vector of the form `[2im 3f0]` or such.  Honestly I don't quite know what the use case is for supporting more general iterables. However,  the [wiki page](https://en.wikipedia.org/wiki/Cauchy_matrix) defines *x_i* and *y_i* to be two "injective sequences" so that language  is more general than vectors.  Perhaps the original code authors were intending hold true to that definition.

The "injective" part of that definition means that each of these two sequences should have unique elements, and currently the code was not testing for that.  I will add a test that reports a warning - let me know what you think.

Reply:
Actually, I've looked back at the history of the original PR by @fp4code
https://github.com/fp4code/SpecialMatrices.jl/blob/e5036efc4cc3c6ece11be6e9c0d90e5b0a9019ae/test/cauchy.jl
and it looks like the most general type being tested was `AbstractArray`, not more general iterables.
So it would be fine with me to simplify this constructor to take only `AbstractArray` types with 1-based indexing.
Would that be preferable to you @mkitti ?

Reply:
I think we should definitely write a method for AbstractVector{T}. We can even leave the older implementation here.